### ISSUE TITLE
Orange Pi 3 LTS img failing to build

### ISSUE THREAD

**Issue by Lenny0939** [katyo/nixos-arm#3]
Running `make` within the nixos-arm directory displays this output:
```
nix-build '<nixpkgs/nixos>' -A config.system.build.sdImage -I nixos-config=./sd-image-opi3lts.nix -o result-opi3lts --option builders '' --argstr system aarch64-linux 2>&1 | tee sd-image-opi3lts.log
trace: warning: system.stateVersion is not set, defaulting to 24.05. Read why this matters on https://nixos.org/manual/nixos/stable/options.html#opt-system.stateVersion.
error:
       … while calling the 'derivationStrict' builtin

         at /builtin/derivation.nix:9:12: (source not available)

       … while evaluating derivation 'nixos-sd-image-sunxi64-orangepi-3-lts-24.05pre554114.e92039b55bcd-aarch64-linux.img'
         whose name attribute is located at /nix/store/ywi1r7f8n81faa09m616a4zqrii4dzyw-nixos/nixos/pkgs/stdenv/generic/make-derivation.nix:348:7

       … while evaluating attribute 'buildCommand' of derivation 'nixos-sd-image-sunxi64-orangepi-3-lts-24.05pre554114.e92039b55bcd-aarch64-linux.img'

         at /home/lenny/Projects/server/nixos-arm/modules/sd-image.nix:244:7:

          243|
          244|       buildCommand = ''
             |       ^
          245|         mkdir -p $out/nix-support $out/sd-image

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: attribute 'version' missing

       at /nix/store/ywi1r7f8n81faa09m616a4zqrii4dzyw-nixos/nixos/nixos/modules/tasks/filesystems/zfs.nix:544:23:

          543|         {
          544|           assertion = cfgZfs.modulePackage.version == cfgZfs.package.version;
             |                       ^
          545|       
```

**Comment by deftdawg** [comment]
I created a pull request with a fix that let's the build start, this error is no longer blocking... 

However for me, qemu-aarch64 locks up at some point during the build process and I'm unable to get an image out of it :cry:

**Comment by Lenny0939** [comment]
Yeah, I got an error too: `error: a 'aarch64-linux' with features {} is required to build '/nix/store/kxr368d7xpgpxvkdax502qlqjvikk7gb-audit-disable.drv', but I am a 'x86_64-linux' with features {benchmark, big-parallel, kvm, nixos-test}`
Building with an arm machine should fix it, hopefully I could somehow emulate it, but if not I could borrow a mac and install Nix on it maybe? I'll research some more and see if it's possible to emulate

**Comment by deftdawg** [comment]
In 2 build attempts, I've gotten locked up at linking sound/soc/codecs/snd-soc-lpass-rx-macro.o

Host is intel 13900k, nixos 24.05/unstable, tried building booted with 6.1 and then 6.7 kernels...
~~~console
/nix/store/wjkn3fmykn68rbpnamj9kzgw263xcga2-binutils-2.40/bin/ld: warning: -z relro ignored
  LD [M]  sound/soc/codecs/snd-soc-wm-adsp.o
/nix/store/wjkn3fmykn68rbpnamj9kzgw263xcga2-binutils-2.40/bin/ld: warning: -z relro ignored
  LD [M]  sound/soc/codecs/snd-soc-wm8994.o
  LD [M]  sound/soc/codecs/snd-soc-lpass-rx-macro.o
~~~

Maybe will see if I can disable sound as I don't really need it for my intended use.

**Comment by Lenchog** [comment]
I can't even get to the audio step, I get stuck at error: `a 'aarch64-linux' with features {} is required to build '/nix/store/ajsa4cdx5l5bxg6z059kk46nzyfy60y0-audit-disable.drv', but I am a 'x86_64-linux' with features {benchmark, big-parallel, kvm, nixos-test}`
Could you build it without audio and send me the file? Or do you know how to fix it? Thanks

**Comment by deftdawg** [comment]
I don’t get that error regarding features, are you using a clean checkout
or did you modify it to add those features?

WRT to my audio problem, I don’t know where to add SOUND=n that it gets
into the kernel config during the build… would be nice if there’s a way to
do it at the top level, but I fear it maybe buried in the overlays… some
code completion for nix files would help a lot 😄

**Comment by Lenchog** [comment]
Yeah, my checkout was clean. I tried building on two different machines, a Ryzen desktop and an Intel laptop, both had the same error. Code completion like that would be nice! Maybe nil-ls + cmp support it? I don’t think it’s that advanced yet though

**Comment by deftdawg** [comment]
Are you on 23.11 or 24.05? (I'm 24.05 because Jovian requires me to be unstable to work properly)

**Comment by Lenchog** [comment]
At least on my laptop, 24.05. Maybe I have disabled some virtualisation feature in my own config? Or need to enable one? I'm picking up my Pi from the post office today so hopefully I can get it working with the generic image, or fix this quickly

**Comment by deftdawg** [comment]
Do you have qemu package?

Do you have this in your configuration.nix?
~~~
  # Build Rpi images
  boot.binfmt.emulatedSystems = [ "aarch64-linux" ];
~~~

**Comment by Lenchog** [comment]
Ohhh, thank you! That must be it! I'll test it out

**Comment by Lenchog** [comment]
By the way, how long does it take to build? On my (relatively) crappy laptop it's been over an hour and it's still compiling the kernel. Is it worth it to swap over to my desktop (R5 3600)?

**Comment by deftdawg** [comment]
No idea, I don't make it past kernel compile... compiling a kernel should be less than 1/2hr I'm guessing.  I always go away and come back and it's stuck 😔 

I've got an i9 13900k w/ 128gb ram, seems to eat about 20gb including cache... on a native kernel it rips through pretty quickly, but emulating arm is obviously gonna slow it down.

**Comment by Lenchog** [comment]
It's been 2 hours now, and still hasn't failed. So I guess that's a good sign. Wow, even on my desktop I only have a ryzen 5 3600 and 16gb, and on my laptop an intel 1235u and 16gb, so I guess it'll take a little while longer.

**Comment by deftdawg** [comment]
Just watch for long pauses...  for me it looks like it's doing something (using 100% cpu), but terminal doesn't update. So when I'm not sure I hit enter 5 or 6 times and type some crap and then check back in a couple of minutes to make sure it's still moving

**Comment by deftdawg** [comment]
Need to figure out how to disable compiling nouveau and amdgpu, those are a huge waste of compiling time 

After about an hour my build died here, though nix-build/qemu is still using 100% on 4 of my cores 😵‍💫 

~~~
  CC [M]  drivers/gpu/drm/nouveau/nv84_fence.o
  CC [M]  drivers/gpu/drm/nouveau/nvc0_fence.o
  LD [M]  drivers/gpu/drm/nouveau/nouveau.o
/nix/store/wjkn3fmykn68rbpnamj9kzgw263xcga2-binutils-2.40/bin/ld: warning: -z relro ignored
make[3]: *** [../scripts/Makefile.build:504: drivers/gpu/drm] Error 2
make[2]: *** [../scripts/Makefile.build:504: drivers/gpu] Error 2
~~~

**Comment by Lenchog** [comment]
Have you tried waiting for it? I've seen the warning about -z relro many times, and hasn't stopped the build

**Comment by deftdawg** [comment]
waited like 5 minutes, it didn't move on to a new task... there's no reason for an error message terminating a compile to take longer than a minute... if was linking or packing a boot image ok, but dying should be quick 😆

Did you get a build? Is it still grinding away on your laptop?

**Comment by Lenchog** [comment]
Yes! It's built! Last night (I'm in Australia) I started compiling on my desktop, and I woke up at about 4:30 (can't help it) and checked, and it seemed to be hung, but I went back to sleep and just checked when I woke up (6:00), and it's complete! My laptops still chugging away lol. Would you like me to send you the image?

**Comment by deftdawg** [comment]
Wow congrats!   Was that built with qemu 8.1.3 or 8.2.0?

Not sure if it would the image would boot on the OPiZero2W, but if you have a Zero2 image I'm game to try it.

On my end I see some of my build problems `journalctl --since="35 minutes ago" | grep "qemu"`:
```log
Jan 23 15:38:02 nixos systemd-coredump[4092261]: Resource limits disable core dumping for process 4091644 (qemu-aarch64).
Jan 23 15:38:02 nixos systemd-coredump[4092261]: Process 4091644 (qemu-aarch64) of user 30001 terminated abnormally without generating a coredump.
Jan 23 15:41:43 nixos systemd-coredump[4147287]: Resource limits disable core dumping for process 4147280 (qemu-aarch64).
Jan 23 15:41:43 nixos systemd-coredump[4147287]: Process 4147280 (qemu-aarch64) of user 30001 terminated abnormally without generating a coredump.
Jan 23 15:50:33 nixos systemd-coredump[87902]: Resource limits disable core dumping for process 86606 (qemu-aarch64).
Jan 23 15:50:33 nixos systemd-coredump[87902]: Process 86606 (qemu-aarch64) of user 30001 terminated abnormally without generating a coredump.
```

But at least I found where the kernel config is:
```patch
diff --git a/overlays/linux/sun50i/default.nix b/overlays/linux/sun50i/default.nix
index 51ec2d3..77171fb 100644
--- a/overlays/linux/sun50i/default.nix
+++ b/overlays/linux/sun50i/default.nix
@@ -159,8 +159,12 @@ let version = "6.2.1";
         DRM_SUN8I_MIXER = module;
         DRM_SUN8I_TCON_TOP = module;
 
+        # Jeff Geerling will have to change this back if he wants to have a go with these.
+        DRM_NOUVEAU = no;
+        DRM_AMDGPU = no;
+
         # H6
-        SOUND = yes; #module;
+        SOUND = no; #module;
         SND_SIMPLE_CARD = module;
         SND_SOC_ACX00 = module;
         SND_SUN4I_I2S = module;
```

**Comment by Lenchog** [comment]
If you want I can compile it tonight (ready in about 20 hours from now), but I'm pretty sure you just need to wait, though we do have different machines and there could be something breaking on yours. I don't know which version of qemu it was, I'm using unstable and I simply ran "nix-shell -p gnumake --run make" to build and didn't change anything, so AFAIK it should be the same as yours.

**Comment by deftdawg** [comment]
All good, it's a bit of a long shot it's going to work on my OPi02w even if I did manage to get it built.  

I think I'll just use dietpi for now (it boots okay)...  that should at least be good enough to test if bridgemode is stable or not on the OrangePi (on my RPi02w, it would hardlock) with my dual nic adapter board..

Perhaps I'll revisit this down the line.  Do let me know if it boots and runs on your OPi though, if not maybe I can offer my troubleshooting 2 cents.  :beers:

**Comment by Lenchog** [comment]
Yep, I got it working! I would like to install it on the EMMC, but I don’t know if the NixOS installation guide will apply, since it’s arm. Do you know how I could potentially get it installed?

**Comment by deftdawg** [comment]
I don't personally have a device with emmc (the zero2w has 16mb, which isn't useful), so don't know...  This page suggests a command for armbian (https://jumptuck.com/blog/2023-02-13-install-linux-orange-pi-3-lts-emmc/), I'd take a look at the script's source to see if it offers clues (https://github.com/speters/armbian-build-lib/blob/master/scripts/nand-sata-install/usr/lib/nand-sata-install/nand-sata-install.sh)...

**Comment by Lenchog** [comment]
That looks complicated! I’m not super experienced with shell and Linux, especially not on arm so I’ll probably just leave it as is for now. I’ve actually run into some trouble with the board, so it was working fine, and then I made a change and couldn’t ssh into it anymore, and when I rebooted it it wouldn’t boot (no green light), so I reinstalled the image, and now the green light comes on but there is no display. I checked and the built in android works fine. Originally I had it plugged it into Ethernet (my monitor and the only Ethernet place in my house are in different rooms, so it’s a bit of a pain to use Ethernet) and also used a different power source, would either of those make a difference?
